fuse-utils/man/tape2wav.1
.\" -*- nroff -*-
.\"
.\" tape2wav.1: tape2wav man page
.\" Copyright (c) 2007 Fredrick Meunier
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.\" Author contact information:
.\"
.\" E-mail: fredm@spamcop.net
.\"
.\"
.TH tape2wav 1 "21st August, 2016" "Version 1.2.2" "Emulators"
.\"
.\"------------------------------------------------------------------
.\"
.SH NAME
tape2wav \(em Sinclair ZX Spectrum tape to wav audio file converter
.\"
.\"------------------------------------------------------------------
.\"
.SH SYNOPSIS
.B tape2wav
.RI "[ \-r " "sample rate" " ]"
.I infile outfile
.\"
.\"------------------------------------------------------------------
.\"
.SH DESCRIPTION
tape2wav converts ZX Spectrum tape images to audio files.
.\"
.\"------------------------------------------------------------------
.\"
.SH OPTIONS
.TP
.RI "\-r " "sample rate"
causes
.I sample rate
to be used as the sample rate for the target wav file. The default rate is
44100 Hz.
.TP
\-h, \-\-help
give brief usage help, listing available options.
.TP
\-V, \-\-version
output version information.
.TP
.I infile
specifies the tape file to be converted. This file can be in any of the
tape formats supported by
.IR libspectrum "(3)."
.TP
.I outfile
specifies the output wav file, if set to \(lq\-\(rq will write the file to standard
output.
.\"
.\"------------------------------------------------------------------
.\"
.SH BUGS
None known.
.\"
.\"------------------------------------------------------------------
.\"
.SH SEE ALSO
.IR fuse "(1),"
.IR fuse\-utils "(1),"
.IR libspectrum "(3),"
.IR listbasic "(1),"
.IR tapeconv "(1),"
.IR tzxlist "(1),"
.IR snap2tzx "(1),"
.IR tape2pulses "(1)"
.IR tape2wav "(1)"
.PP
The comp.sys.sinclair Spectrum FAQ, at
.br
.IR "http://www.worldofspectrum.org/faq/index.html" .
.\"
.\"------------------------------------------------------------------
.\"
.SH AUTHOR
Fredrick Meunier (fredm@spamcop.net).


fuse/ChangeLog
2016-07-17  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * Fuse 1.2.1 released.

        * Emulation core improvements:
          * Fix bugs when the detect loaders feature is being used (thanks, Lee
            Tonks and windale) (Fredrick Meunier).

        * Debugger improvements:
          * Remove the need for "%" when accessing system variables (Philip
            Kendall).
          * Add Z80 registers as debugger variables (Philip Kendall).
          * Expose last byte written to the ULA, tstates since interrupt,
            primary and secondary memory control ports as debugger system
            variables (Philip Kendall).
          * Make breakpoints on events honour lifetime (Sergio Baldoví).
          * Extend breakpoints on paging events to more peripherals: Beta 128,
            +D, Didaktik 80, DISCiPLE, Opus Discovery and SpeccyBoot (Sergio
            Baldoví).
          * Split +D memory sources into RAM and ROM sections (Sergio Baldoví).
          * Coalesce +D and DISCiPLE RAM pages so they show as 8K pages (Sergio
            Baldoví).

        * Miscellaneous improvements:
          * Update GNOME .desktop file (Alberto Garcia).
          * Add an emulator module startup manager to automatically handle
            dependency issues (Philip Kendall).
          * Fix crash on widget UIs when hitting the close icon on the title bar
            several times (Sergio Baldoví).

2016-06-06  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * Fuse 1.2.0 released.

        * New features:
          * Add Currah µSource emulation (Stuart Brady).
          * Add Didaktik 80/40 emulation (Gergely Szasz).
          * Capture BASIC video functions to SVG file (Stefano Bodrato).
          * Support bash completion (Sergio Baldoví).

        * Emulation core improvements:
          * Allow continuing RZX recordings if there is a final snapshot in the
            RZX (Sergio Baldoví).
          * Fix the prune function on RZX rollback (Sergio Baldoví).
          * Use SZX format for the initial snapshot in RZX files (Sergio
            Baldoví).
          * Fix loading of EDSK files with Sector Offset block (Sergio Baldoví).
          * Migrate disk "index event" handling to the FDD layer and have the
            FDC layer use it for their STATUS registers (Gergely Szasz).
          * Implement WD2797 emulation (Gergely Szasz).
          * Centralise the "Disk icon" update code to the FDD layer (Gergely
            Szasz).
          * Fix disk image corruption after saving UDI files (Sergio Baldoví).
          * Check ready status after loading a disk into floppy disk drive
            (thanks, John Elliott) (Sergio Baldoví).
          * Fix overlapped SEEK commands (Sergio Baldoví).
          * Fix length of data returned by READ_DIAG (thanks, Fredrick Meunier)
            (Sergio Baldoví).
          * Fix writing .td0 format disk files (Sergio Baldoví).
          * Fix speech in Cobra's Arc - Medium Case.tzx when loaded with tape
            traps enabled (thanks, zx81 and Sergio Baldoví) (Fredrick Meunier).
          * Skip tape traps if VERIFY is requested (UB880D).
          * Fix loading sound with some custom loaders (Fredrick Meunier).
          * Check if data blocks are headers for handling PZX files (Fredrick
            Meunier).
          * Correct display of pulses in PZX pulse block for tape browser
            (Fredrick Meunier).
          * Set AF, AF' and SP to 0xffff on reset (Stuart Brady).
          * Leave most registers unchanged on a soft reset (Stuart Brady).
          * Emulate interrupt and NMI timings more precisely (Stuart Brady).
          * Emulate NMOS and CMOS Z80 variants (Stuart Brady).
          * Switch to 2KB page size (Stuart Brady).
          * Fix inaccurate output when AY envelopes are used (Matthew Westcott
            and Fredrick Meunier).
          * Fix some peripherals activation when loading snapshots (Sergio
            Baldoví).

        * Machine specific improvements:
          * The address range 0x4000 - 0x7FFF is contended on the TS2068 in the
            home, Dock and Exrom banks (thanks, Richard Atkinson). It is assumed
            that this is the same for other Timex models (Fredrick Meunier).
          * Clear all Opus RAM on hard reset (Stuart Brady).
          * Ensure the ZX Printer does not require a hard reset to enable
            (thanks, RMartins) (Fredrick Meunier).
          * Fix timing of events when emulating Scorpion (Stuart Brady).
          * Ensure we have successfully selected a Pentagon or a Timex machine
            before inserting their media (Fredrick Meunier).
          * Prefer Scorpion to Pentagon when loading SCL/TRD disks for better
            timing compatibility (thanks, windale) (Sergio Baldoví).
          * Fix Beta 128 type II commands (thanks, windale and BogDan Vatra)
            (Sergio Baldoví).
          * Don't enable the Beta 128 interface when loading a snapshot on a
            machine with Beta built-in (thanks, windale and BogDan Vatra)
            (Fredrick Meunier).
          * Lock port +3 1FFDh if paging is disabled (Brian Ruthven and Fredrick
            Meunier).
          * Spectranet: fix segfault in error handling when setting SO_REUSEADDR
            (Stuart Brady).
          * Update SE ROMs to v4.07 (thanks, Andrew Owen) (Sergio Baldoví).

        * Debugger improvements:
          * Fix disassembly of LD (HL), LD (IX) and LD (IY) (BogDan Vatra).
          * Add I and R register setting and getting to the debugger (Sergio
            Baldoví).
          * Show the status of the halted flag in the debugger (Stuart Brady).
          * Fix time breakpoints later than a frame in the future (Sergio
            Baldoví).
          * Timex EXROM and Dock text was truncated in the debugger UI (thanks,
            Andrew Owen and Sergio Baldoví) (Fredrick Meunier).
          * Win32: relocate halted flag and set monospaced font (Sergio

fuse-utils/tape2wav.c
/* tape2wav.c: Convert tape files (tzx, tap, etc.) to .wav files
   Copyright (c) 2007 Fredrick Meunier
   Copyright (c) 2014-2015 Sergio Baldovi

   $Id$

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

   Author contact information:

   E-mail: fredm@spamcop.net

*/

#include <config.h>

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include <libspectrum.h>

#include <audiofile.h>

#include "utils.h"

#define PROGRAM_NAME "tape2wav"

static void show_help( void );
static void show_version( void );
static int read_tape( char *filename, libspectrum_tape **tape );
static int write_tape( char *filename, libspectrum_tape *tape );

char *progname;
int sample_rate = 44100;

int
main( int argc, char **argv )
{
  int c, error = 0;
  libspectrum_tape *tzx;

  progname = argv[0];

  struct option long_options[] = {
    { "help", 0, NULL, 'h' },
    { "version", 0, NULL, 'V' },
    { 0, 0, 0, 0 }
  };

  while( ( c = getopt_long( argc, argv, "r:hV", long_options, NULL ) ) != -1 ) {

    switch( c ) {

    case 'r': sample_rate = abs( atoi( optarg ) ); break;
    case 'h': show_help(); return 0;
    case 'V': show_version(); return 0;

    case '?':
      /* getopt prints an error message to stderr */
      error = 1;
      break;

    default:
      error = 1;
      fprintf( stderr, "%s: unknown option `%c'\n", progname, (char) c );
      break;

    }

  }

  argc -= optind;
  argv += optind;

  if( error ) {
    fprintf( stderr, "Try `%s --help' for more information.\n", progname );
    return error;
  }

  if( argc < 2 ) {
    fprintf( stderr,
             "%s: usage: %s [-r rate] <infile> <outfile>\n",
             progname,
	     progname );
    fprintf( stderr, "Try `%s --help' for more information.\n", progname );
    return 1;
  }

  error = init_libspectrum(); if( error ) return error;

  if( read_tape( argv[0], &tzx ) ) return 1;

  if( write_tape( argv[1], tzx ) ) {
    libspectrum_tape_free( tzx );
    return 1;
  }

  libspectrum_tape_free( tzx );

  return 0;
}

static void
show_version( void )
{
  printf(
    PROGRAM_NAME " (" PACKAGE ") " PACKAGE_VERSION "\n"
    "Copyright (c) 2007 Fredrick Meunier\n"
    "License GPLv2+: GNU GPL version 2 or later "
    "<http://gnu.org/licenses/gpl.html>\n"
    "This is free software: you are free to change and redistribute it.\n"
    "There is NO WARRANTY, to the extent permitted by law.\n" );
}

static void
show_help( void )
{
  printf(
    "Usage: %s [OPTION] <infile> <outfile>\n"
    "Converts ZX Spectrum tape images to audio files.\n"
    "\n"
    "Options:\n"
    "  -r <sample rate>  Set the sample rate for the target wav file. Defaults to\n"
    "                    44100 Hz.\n"
    "  -h, --help     Display this help and exit.\n"
    "  -V, --version  Output version information and exit.\n"
    "\n"
    "Report %s bugs to <%s>\n"
    "%s home page: <%s>\n"
    "For complete documentation, see the manual page of %s.\n",
    progname,
    PROGRAM_NAME, PACKAGE_BUGREPORT, PACKAGE_NAME, PACKAGE_URL, PROGRAM_NAME
  );
}

static int
read_tape( char *filename, libspectrum_tape **tape )
{
  libspectrum_byte *buffer; size_t length;

  if( read_file( filename, &buffer, &length ) ) return 1;

  *tape = libspectrum_tape_alloc();

  if( libspectrum_tape_read( *tape, buffer, length, LIBSPECTRUM_ID_UNKNOWN,
                             filename ) ) {
    free( buffer );
    return 1;
  }

  free( buffer );

  return 0;
}

static int
write_tape( char *filename, libspectrum_tape *tape )
{
  libspectrum_byte *buffer, *ptr; size_t length;
  size_t tape_length = 0;
  libspectrum_error error;
  short level = 0; /* The last level output to this block */
  libspectrum_dword pulse_tstates = 0;
  libspectrum_dword balance_tstates = 0;
  int flags = 0;
  int framesWritten;
  AFfilehandle file;

  AFfilesetup setup = afNewFileSetup();

  unsigned int scale = 3500000/sample_rate;

  length = 8192;
  buffer = malloc( length );

  if( !buffer ) {
    fprintf( stderr, "%s: unable to allocate memory for conversion buffer\n",
             progname );
    return 1;
  }

  while( !(flags & LIBSPECTRUM_TAPE_FLAGS_TAPE) ) {
    libspectrum_dword pulse_length = 0;
    size_t i;

    error = libspectrum_tape_get_next_edge( &pulse_tstates, &flags, tape );
    if( error != LIBSPECTRUM_ERROR_NONE ) {
      free( buffer );
      return 1;
    }

    /* Invert the microphone state */
    if( pulse_tstates ||
        !( flags & LIBSPECTRUM_TAPE_FLAGS_NO_EDGE ) ||
        ( flags & ( LIBSPECTRUM_TAPE_FLAGS_STOP |
                    LIBSPECTRUM_TAPE_FLAGS_LEVEL_LOW |
                    LIBSPECTRUM_TAPE_FLAGS_LEVEL_HIGH ) ) ) {

      if( flags & LIBSPECTRUM_TAPE_FLAGS_NO_EDGE ) {
        /* Do nothing */
      } else if( flags & LIBSPECTRUM_TAPE_FLAGS_LEVEL_LOW ) {
        level = 0;
      } else if( flags & LIBSPECTRUM_TAPE_FLAGS_LEVEL_HIGH ) {
        level = 1;
      } else {
        level = !level;
      }

    }

    balance_tstates += pulse_tstates;

    if( flags & LIBSPECTRUM_TAPE_FLAGS_NO_EDGE ) continue;

    pulse_length = balance_tstates / scale;
    balance_tstates = balance_tstates % scale;

    /* TZXs produced by snap2tzx have very tight tolerances, err on the side of
       producing a pulse that is too long rather than too short */
    if( balance_tstates > scale>>1 ) {
      pulse_length++;
      balance_tstates = 0;
    }

    while( tape_length + pulse_length > length ) {
      length *= 2;
      ptr = buffer;
      buffer = realloc( buffer, length );
      if( !buffer ) {
        free( ptr );
        fprintf( stderr, "%s: unable to allocate memory for conversion buffer\n",
                 progname );
        return 1;
      }
    }

    for( i = 0; i < pulse_length; i++ ) {
      buffer[ tape_length++ ] = level ? 0xff : 0x00;
    }
  }

  afInitFileFormat( setup, AF_FILE_WAVE );
  afInitChannels( setup, AF_DEFAULT_TRACK, 1 );
  afInitSampleFormat( setup, AF_DEFAULT_TRACK, AF_SAMPFMT_UNSIGNED, 8 );
  afInitRate( setup, AF_DEFAULT_TRACK, sample_rate );

  if( strncmp( filename, "-", 1 ) == 0 ) {
    int fd = fileno( stdout );
    if( isatty( fd ) ) {
      fprintf( stderr, "%s: won't output binary data to a terminal\n",
               progname );
      free( buffer );
      afFreeFileSetup( setup );
      return 1;
    }

#ifdef WIN32
    setmode( fd, O_BINARY );
#endif				/* #ifdef WIN32 */

    file = afOpenFD( fd, "w", setup );
  } else {
    file = afOpenFile( filename, "w", setup );
  }
  if( file == AF_NULL_FILEHANDLE ) {
    fprintf( stderr, "%s: unable to open file '%s' for writing\n", progname,
             filename );
    free( buffer );
    afFreeFileSetup( setup );
    return 1;
  }

  framesWritten = afWriteFrames( file, AF_DEFAULT_TRACK, buffer, tape_length );
  if(framesWritten != tape_length ) {
    fprintf( stderr,
             "%s: number of frames written does not match number of frames"
             " requested\n",
             progname );
    free( buffer );
    afFreeFileSetup( setup );
    return 1;
  }

  if( afCloseFile( file ) ) {
    fprintf( stderr, "%s: error closing wav file\n", progname );
    free( buffer );
    afFreeFileSetup( setup );
    return 1;
  }

  free( buffer );
  afFreeFileSetup( setup );

  return 0;
}
